File: oxidize-python/oxidize_python/core/tensor/dtype.py
"""DType and Tensor mirroring oxidize-golang/core/tensor/dtype.go."""

from __future__ import annotations

import struct
import threading
from dataclasses import dataclass
from enum import IntEnum, auto


class DType(IntEnum):
    F32 = 0
    F16 = 1
    I8 = 2
    I16 = 3
    I32 = 4
    I64 = 5

    def size_in_bytes(self) -> int:
        match self:
            case DType.F32 | DType.I32:
                return 4
            case DType.F16 | DType.I16:
                return 2
            case DType.I8:
                return 1
            case DType.I64:
                return 8
        return 0

    def __str__(self) -> str:
        return self.name.lower()


class ActivationFn(IntEnum):
    NONE = auto()
    SILU = auto()
    GELU = auto()
    RELU = auto()


@dataclass
class Tensor:
    _mu: threading.RLock
    _data: list[float]
    _bytes: bytes | bytearray
    shape: list[int]
    strides: list[int]
    dtype: DType
    owned: bool

    @classmethod
    def new(cls, data: list[float], shape: list[int]) -> Tensor | None:
        total = 1
        for s in shape:
            if s <= 0:
                return None
            total *= s
        if total != len(data):
            return None
        t = cls(
            _mu=threading.RLock(),
            _data=list(data),
            _bytes=b"",
            shape=list(shape),
            strides=_default_strides(shape),
            dtype=DType.F32,
            owned=True,
        )
        return t

    @classmethod
    def from_bytes(cls, data: bytes | bytearray, shape: list[int], dt: DType) -> Tensor:
        total = 1
        for s in shape:
            if s <= 0:
                raise ValueError(f"invalid shape dimension: {s}")
            total *= s
        if total * dt.size_in_bytes() != len(data):
            raise ValueError(f"shape/dtype mismatch: shape={shape} dtype={dt} bytes={len(data)}")
        f32 = list(_bytes_as_f32(data)) if dt == DType.F32 else []
        return cls(
            _mu=threading.RLock(),
            _data=f32,
            _bytes=data,
            shape=list(shape),
            strides=_default_strides(shape),
            dtype=dt,
            owned=dt == DType.F32,
        )

    def data(self) -> list[float]:
        with self._mu:
            if self.dtype == DType.F32:
                return list(self._data)
            n = self.num_elements()
            return [self._f32_at(i) for i in range(n)]

    def bytes(self) -> bytes:
        with self._mu:
            if self.dtype == DType.F32:
                buf = bytearray(len(self._data) * 4)
                for i, v in enumerate(self._data):
                    struct.pack_into("<f", buf, i * 4, v)
                return bytes(buf)
            return bytes(self._bytes)

    def num_elements(self) -> int:
        n = 1
        for s in self.shape:
            n *= s
        return n

    def at(self, i: int) -> float:
        with self._mu:
            return self._f32_at(i)

    def set(self, i: int, v: float) -> None:
        with self._mu:
            match self.dtype:
                case DType.F32:
                    self._data[i] = v
                case DType.F16:
                    bits = f32_to_f16_bits(v)
                    self._bytes[i * 2] = bits & 0xFF
                    self._bytes[i * 2 + 1] = (bits >> 8) & 0xFF
                case DType.I8:
                    self._bytes[i] = int(max(-128, min(127, int(v)))) & 0xFF
                case DType.I16:
                    struct.pack_into("<h", self._bytes, i * 2, int(max(-32768, min(32767, int(v)))))
                case DType.I32:
                    struct.pack_into("<i", self._bytes, i * 4, int(v))
                case DType.I64:
                    struct.pack_into("<q", self._bytes, i * 8, int(v))

    def _f32_at(self, i: int) -> float:
        match self.dtype:
            case DType.F32:
                return self._data[i]
            case DType.F16:
                bits = self._bytes[i * 2] | (self._bytes[i * 2 + 1] << 8)
                return f16_bits_to_f32(bits)
            case DType.I8:
                b = self._bytes[i]
                return float(b - 256 if b > 127 else b)
            case DType.I16:
                return float(struct.unpack_from("<h", self._bytes, i * 2)[0])
            case DType.I32:
                return float(struct.unpack_from("<i", self._bytes, i * 4)[0])
            case DType.I64:
                return float(struct.unpack_from("<q", self._bytes, i * 8)[0])
        return 0.0


def _default_strides(shape: list[int]) -> list[int]:
    strides = [0] * len(shape)
    if not shape:
        return strides
    stride = 1
    for i in range(len(shape) - 1, -1, -1):
        strides[i] = stride
        stride *= shape[i]
    return strides


def _bytes_as_f32(b: bytes | bytearray) -> list[float]:
    return [struct.unpack_from("<f", b, i * 4)[0] for i in range(len(b) // 4)]


def f16_le_to_f32(b: tuple[int, int] | bytes) -> float:
    if isinstance(b, bytes):
        bits = b[0] | (b[1] << 8)
    else:
        bits = b[0] | (b[1] << 8)
    return f16_bits_to_f32(bits)


def f16_bits_to_f32(bits: int) -> float:
    sign = (bits >> 15) & 1
    exp = (bits >> 10) & 0x1F
    mant = bits & 0x3FF
    if exp == 0:
        if mant == 0:
            out = sign << 31
        else:
            while mant & 0x400 == 0:
                mant <<= 1
                exp -= 1
            exp += 1
            mant &= 0x3FF
            out = (sign << 31) | ((exp + 112) << 23) | (mant << 13)
    elif exp == 0x1F:
        out = (sign << 31) | (0xFF << 23) | (mant << 13)
    else:
        out = (sign << 31) | ((exp + 112) << 23) | (mant << 13)
    return struct.unpack("<f", struct.pack("<I", out))[0]


def f32_to_f16_bits(f: float) -> int:
    bits = struct.unpack("<I", struct.pack("<f", f))[0]
    sign = (bits >> 31) & 1
    exp = ((bits >> 23) & 0xFF) - 127
    mant = bits & 0x7FFFFF
    if exp == 128:
        return (sign << 15) | (0x7C00 if mant == 0 else 0x7C00 | (mant >> 13))
    if exp > 15:
        return (sign << 15) | 0x7C00
    if exp < -24:
        return sign << 15
    if exp < -13:
        mant |= 0x800000
        shift = -10 - exp
        mant = (mant + (1 << (shift - 1))) >> shift
        return (sign << 15) | mant
    return (sign << 15) | ((exp + 15) << 10) | (mant >> 13)


def extract_bits(bitstream: bytes | bytearray, index: int, bits: int) -> int:
    if bits == 0 or bits > 32:
        return 0
    bit_pos = index * bits
    byte_pos = bit_pos // 8
    bit_offset = bit_pos % 8
    value = 0
    remaining = bits
    shift = 0
    while remaining > 0:
        available = 8 - bit_offset
        take = min(remaining, available)
        mask = (1 << take) - 1
        chunk = (bitstream[byte_pos] >> bit_offset) & mask
        value |= chunk << shift
        remaining -= take
        shift += take
        bit_offset = 0
        byte_pos += 1
    return value

File: oxidize-golang/core/tensor/dtype.go
// Package tensor mirrors oxidize_core::compute::tensor. It provides DType,
package tensor

import (
	"encoding/binary"
	"fmt"
	"math"
	"sync"
)

// DType identifies the element type of a tensor.
type DType uint8

const (
	DTypeF32 DType = iota
	DTypeF16
	DTypeI8
	DTypeI16
	DTypeI32
	DTypeI64
)

func (d DType) String() string {
	switch d {
	case DTypeF32:
		return "f32"
	case DTypeF16:
		return "f16"
	case DTypeI8:
		return "i8"
	case DTypeI16:
		return "i16"
	case DTypeI32:
		return "i32"
	case DTypeI64:
		return "i64"
	default:
		return fmt.Sprintf("dtype(%d)", uint8(d))
	}
}

// SizeInBytes returns the byte width of a single element of the dtype.
func (d DType) SizeInBytes() int {
	switch d {
	case DTypeF32, DTypeI32:
		return 4
	case DTypeF16, DTypeI16:
		return 2
	case DTypeI8:
		return 1
	case DTypeI64:
		return 8
	default:
		return 0
	}
}

// ActivationFn identifies a hidden activation function for linear layers.
type ActivationFn uint8

const (
	ActivationNone ActivationFn = iota
	ActivationSilu
	ActivationGelu
	ActivationRelu
)

func (a ActivationFn) String() string {
	switch a {
	case ActivationSilu:
		return "silu"
	case ActivationGelu:
		return "gelu"
	case ActivationRelu:
		return "relu"
	default:
		return "none"
	}
}

// Kernel-related constants.
const (
	FlashAttentionBlockTokens  = 64
	ParallelGemvMinOps         = 4096
	TransposedGemvColChunk     = 64
)

// Common block sizes re-exported for convenience.
const (
	QK8_0     = 32
	QK4_0     = 32
	QK4_1     = 32
	QK5_0     = 32
	QK5_1     = 32
	QK_K      = 256
	QK_NVFP4  = 64
	QK_NVFP4_SUB = 16
)

// Tensor is a row-major, dense, n-dimensional array. The implementation
// mirrors the bottom-up Tensor struct in oxidize-core/src/compute/tensor.rs
// (constants, errors, kernels, struct) using a single struct with helpers.
type Tensor struct {
	mu       sync.RWMutex
	data     []float32
	bytes    []byte
	shape    []int
	strides  []int
	dtype    DType
	owned    bool
}

// New constructs a Tensor from a float32 slice and shape. The data is copied.
func New(data []float32, shape []int) *Tensor {
	t := &Tensor{shape: append([]int(nil), shape...), dtype: DTypeF32, owned: true}
	total := 1
	for _, s := range shape {
		if s <= 0 {
			return nil
		}
		total *= s
	}
	if total != len(data) {
		return nil
	}
	t.data = append([]float32(nil), data...)
	t.strides = defaultStrides(shape)
	return t
}

// FromBytes constructs a Tensor from a raw byte buffer with the given shape
// and dtype. The bytes are referenced (not copied) when copyBytes is false.
func FromBytes(data []byte, shape []int, dt DType) (*Tensor, error) {
	total := 1
	for _, s := range shape {
		if s <= 0 {
			return nil, fmt.Errorf("invalid shape dimension: %d", s)
		}
		total *= s
	}
	if total*dt.SizeInBytes() != len(data) {
		return nil, fmt.Errorf("shape/dtype mismatch: shape=%v dtype=%v bytes=%d", shape, dt, len(data))
	}
	t := &Tensor{bytes: data, shape: append([]int(nil), shape...), dtype: dt, owned: false, strides: defaultStrides(shape)}
	if dt == DTypeF32 {
		t.data = bytesAsF32(data)
		t.owned = true
	}
	return t, nil
}

func defaultStrides(shape []int) []int {
	strides := make([]int, len(shape))
	if len(shape) == 0 {
		return strides
	}
	stride := 1
	for i := len(shape) - 1; i >= 0; i-- {
		strides[i] = stride
		stride *= shape[i]
	}
	return strides
}

// Data returns a copy of the underlying float32 data.
func (t *Tensor) Data() []float32 {
	t.mu.RLock()
	defer t.mu.RUnlock()
	if t.dtype == DTypeF32 {
		return append([]float32(nil), t.data...)
	}
	out := make([]float32, t.numElements())
	for i := range out {
		out[i] = t.f32At(i)
	}
	return out
}

// Bytes returns a copy of the underlying raw bytes.
func (t *Tensor) Bytes() []byte {
	t.mu.RLock()
	defer t.mu.RUnlock()
	if t.dtype == DTypeF32 {
		buf := make([]byte, len(t.data)*4)
		for i, v := range t.data {
			binary.LittleEndian.PutUint32(buf[i*4:], math.Float32bits(v))
		}
		return buf
	}
	return append([]byte(nil), t.bytes...)
}

// Shape returns the shape of the tensor.
func (t *Tensor) Shape() []int { return append([]int(nil), t.shape...) }

// DType returns the dtype of the tensor.
func (t *Tensor) DType() DType { return t.dtype }

// NumElements returns the total element count.
func (t *Tensor) NumElements() int { return t.numElements() }

func (t *Tensor) numElements() int {
	n := 1
	for _, s := range t.shape {
		n *= s
	}
	return n
}

// At returns the float32 value at the given linear offset. For f16/i8 tensors
// it converts on the fly.
func (t *Tensor) At(i int) float32 {
	t.mu.RLock()
	defer t.mu.RUnlock()
	return t.f32At(i)
}

func (t *Tensor) f32At(i int) float32 {
	switch t.dtype {
	case DTypeF32:
		return t.data[i]
	case DTypeF16:
		return F16BitsToF32(uint16(t.bytes[i*2]) | uint16(t.bytes[i*2+1])<<8)
	case DTypeI8:
		return float32(int8(t.bytes[i]))
	case DTypeI16:
		v := int16(binary.LittleEndian.Uint16(t.bytes[i*2:]))
		return float32(v)
	case DTypeI32:
		v := int32(binary.LittleEndian.Uint32(t.bytes[i*4:]))
		return float32(v)
	case DTypeI64:
		v := int64(binary.LittleEndian.Uint64(t.bytes[i*8:]))
		return float32(v)
	default:
		return 0
	}
}

// Set writes a float32 value at the given linear offset. Conversion to the
// underlying dtype is performed as needed.
func (t *Tensor) Set(i int, v float32) {
	t.mu.Lock()
	defer t.mu.Unlock()
	switch t.dtype {
	case DTypeF32:
		t.data[i] = v
	case DTypeF16:
		bits := F32ToF16Bits(v)
		t.bytes[i*2] = byte(bits)
		t.bytes[i*2+1] = byte(bits >> 8)
	case DTypeI8:
		t.bytes[i] = byte(int8(clampInt(v, -128, 127)))
	case DTypeI16:
		binary.LittleEndian.PutUint16(t.bytes[i*2:], uint16(int16(clampInt(v, -32768, 32767))))
	case DTypeI32:
		binary.LittleEndian.PutUint32(t.bytes[i*4:], uint32(int32(v)))
	case DTypeI64:
		binary.LittleEndian.PutUint64(t.bytes[i*8:], uint64(int64(v)))
	}
}

func bytesAsF32(b []byte) []float32 {
	out := make([]float32, len(b)/4)
	for i := range out {
		out[i] = math.Float32frombits(binary.LittleEndian.Uint32(b[i*4:]))
	}
	return out
}

// F16LEToF32 converts a little-endian f16 byte pair to f32.
func F16LEToF32(b [2]byte) float32 {
	bits := uint16(b[0]) | uint16(b[1])<<8
	return F16BitsToF32(bits)
}

// F16BitsToF32 converts a 16-bit IEEE-754 half-precision value to float32.
func F16BitsToF32(bits uint16) float32 {
	sign := uint32(bits>>15) & 0x1
	exp := uint32(bits>>10) & 0x1F
	mant := uint32(bits) & 0x3FF
	var out uint32
	switch {
	case exp == 0:
		if mant == 0 {
			out = sign << 31
		} else {
			// subnormal
			for mant&0x400 == 0 {
				mant <<= 1
				exp--
			}
			exp++
			mant &= 0x3FF
			out = (sign << 31) | ((exp + 112) << 23) | (mant << 13)
		}
	case exp == 0x1F:
		out = (sign << 31) | (0xFF << 23) | (mant << 13)
	default:
		out = (sign << 31) | ((exp + 112) << 23) | (mant << 13)
	}
	return math.Float32frombits(out)
}

// F32ToF16Bits converts a float32 to a 16-bit IEEE-754 half-precision value.
func F32ToF16Bits(f float32) uint16 {
	bits := math.Float32bits(f)
	sign := uint16((bits >> 31) & 0x1)
	exp := int32((bits>>23)&0xFF) - 127
	mant := bits & 0x7FFFFF
	var out uint16
	switch {
	case exp == 128:
		// inf/nan
		if mant == 0 {
			out = (sign << 15) | 0x7C00
		} else {
			out = (sign << 15) | 0x7C00 | uint16(mant>>13)
		}
	case exp > 15:
		out = (sign << 15) | 0x7C00
	case exp < -24:
		out = sign << 15
	default:
		if exp < -13 {
			mant |= 0x800000
			shift := uint(-10 - exp)
			mant = (mant + (1 << (shift - 1))) >> shift
			out = (sign << 15) | uint16(mant)
		} else {
			out = (sign << 15) | uint16(exp+15)<<10 | uint16(mant>>13)
		}
	}
	return out
}

// ExtractBits pulls a `bits`-wide field out of `bitstream` starting at the
// given bit index.
func ExtractBits(bitstream []byte, index int, bits int) uint32 {
	if bits == 0 || bits > 32 {
		return 0
	}
	bitPos := uint(index * bits)
	bytePos := bitPos / 8
	bitOffset := bitPos % 8
	var value uint32
	remaining := bits
	shift := 0
	for remaining > 0 {
		available := 8 - int(bitOffset)
		take := available
		if remaining < take {
			take = remaining
		}
		mask := uint32((1 << take) - 1)
		chunk := uint32(bitstream[bytePos]) >> bitOffset
		value |= (chunk & mask) << shift
		remaining -= take
		shift += take
		bitOffset = 0
		bytePos++
	}
	return value
}

func clampInt(v float32, lo, hi int64) int64 {
	iv := int64(v)
	if iv < lo {
		return lo
	}
	if iv > hi {
		return hi
	}
	return iv
}
